rfcs/0131-sri-hashes.md

```diff
@@ -0,0 +1,50 @@
+---
+feature: sri-hashes
+start-date: 2022-07-25
+author: Winter
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Use SRI hashes and the `hash` parameter for fetchers.
+
+# Motivation
+[motivation]: #motivation
+
+Currently in Nixpkgs, the hash formats that are used in parameters to fetchers range from SRI and base-32 (the most common ones) to hex. In addition to this, packages use the algorithm-specific arguments (`sha256`, `sha512`, etc.) or the generic `hash` argument at random. This creates inconsistencies that we'd rather not have.
+
+# Detailed design
+[design]: #detailed-design
+
+1. Require all new packages to use SRI hashes + `hash` when possible
+2. Do a treewide migration for the majority of cases (single fetcher in `src`, using a fetcher that supports `hash` and SRI hashes (`fetchFromGitHub`, `fetchurl`, etc.)) to `hash` + SRI hashes
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None at this time.
```

Comment:
There's (anecdotically) still a significant number of Nix 2.3 users who contributing to nixpkgs would be made unnecessarily harder.

Reply:
Oh no, another ten years of deprecation window...

Reply:
> Why? This is not a motivation, just a claim.

One of the reasons might be `sha*` are semantically mutually exclusive, but not syntactically – you can specify both `sha256` and `sha512` at the same time by accident (pretty easy when using an override), but you cannot specify multiple different `hash` attributes.

Reply:
You joke, but I seriously think this would be a sensible thing to do.

Reply:
Overall we have developed a strange culture of removing stuff unnecessarily in nixpkgs, I feel like.

In this case, we shouldn't remove support for these formats from Nix anyways (to continue to be able to evaluate older nixpkgs revisions (the ability to reproduce *ancient* stuff truely is a strength of Nix)) and for existing APIs in nixpkgs it's laughably simple to keep supporting it (you don't need to change anything, in fact!). For all I care, stop supporting the `sha256` and `sha512` attributes for new APIs…

Reply:
The text does not talk about removal, I suppose, but about enforcement. Older expressions can be rendered by newer versions, whereas new expressions should be written in newer versions of Nix.

Reply:
Nix 2.3 users ARE relevant because it makes contributing harder, since hashes need to be converted manually.

Reply:
> The text does not talk about removal, I suppose, but about enforcement.

you can only "enforce" your policy by "removing" the alternative

all software has [legacy parts](https://en.wikipedia.org/wiki/Legacy_system) for backward-compatibility.
with nix we will always have `sha256 = "..."`

related:

> the first rule of kernel maintenance:
> [WE DO NOT BREAK USERSPACE!](https://linuxreviews.org/WE_DO_NOT_BREAK_USERSPACE)

Reply:
> Nix 2.3 users ARE relevant because it makes contributing harder, since hashes need to be converted manually.

Is this so hard to the point of being utterly impossible to copy the output of a command?

Okay then: we can relegate it to some form of merge bot.

Reply:
> you can only "enforce" your policy by "removing" the alternative

Please elaborate more than scary quotes.

> with nix we will always have stuff like

Do not conflate nix with nixpkgs.

Anyone can change such hashes to SRI if they wish - both inside and outside Nixpkgs -, and this will not affect Nix 2.3 users.
Indeed no one of the anecdotal users of Nix 2.3 complained about being unable to run the newest KDE Plasma from Nixpkgs because of our new hash format.

> the first rule of kernel maintenance: WE DO NOT BREAK USERSPACE!

Which user space will be broken by this?

Reply:
Removal of support from fetchers would surely break various development environments (outside nixpkgs itself).  But that's undecided in the current proposal.

Reply:
Because of it I am not for removal.
I just think that some things like "why fetchurl accepts hash but fetchgit does not accept" should be resolved and uniformized.

Reply:
> Is this so hard to the point of being utterly impossible to copy the output of a command?

It's not *hard*, just annoying – making the already annoying task of updating hashes even more annoying. Also it will surely mean more round trip in reviews when people unaware of this policy submit PRs with the wrong hash format.

I really don't see the benefit of such a change that justifies wasting everyone's time.

Reply:
> making the already annoying task of updating hashes even more annoying

Annoying tasks should be amenable to automation - like the [r-ryantm](https://github.com/r-ryantm) bot that already uses SRI hashes.

> more round trip in reviews when people unaware of this policy

"People unaware of policies" is a common thing around Nixpkgs. Have you ever seen how many new people want to contribute with new packages? This is very annoying to fix their errors that could be easily avoided by reading the documentation.

Wrong `meta` attributes, wrong version strings, convoluted expressions trying to express unusual building instructions...

Reply:
>> "People unaware of policies" is a common thing around Nixpkgs.

> We should limit ourselves to policies that are actually worth enforcing to keep the friction caused by this problem in check.

Or we should make sure policies are very explicitly and clearly written and ideally automatically checked. The state of the art is that every reviewer has his own set of policies that are unknown beforehand and inconsistent. That's in my experience what causes most back-and-forth delay in the process. Although I accept the "per reviewer policies" are usually not about hash format

Reply:
Not even, there are a few clearly stated policies stated in CONTRIBUTING.md which are routinely ignored by contributors.

Reply:
> Indeed no one of the anecdotal users of Nix 2.3 complained about being unable to run the newest KDE Plasma from Nixpkgs because of our new hash format.

How is this relevant? The claim is that Nix 2.3 can evaluate SRI hashes, but makes updating them annoying; that Nix 2.3 can evaluate them does not refute the claim.

Reply:
> How is this relevant?

To the point someone suggested that it will break the userspace, and, well, no userspace was broken by the new hash format in a long heck of time.

Follow the flow of argument being made.

> The claim is that Nix 2.3 can evaluate SRI hashes, but makes updating them annoying; 

I was not arguing this point, but the "muh userspace breaking" point.

Follow the flow of argument being made.

Reply:
This thread started with the claim that requiring SRI hash format makes _contributing_ while using Nix 2.3 harder (because it doesn't default to printing them), not about _evaluating_ Nixpkgs. The userspace-breaking argument was that we won't remove non-SRI hash support from _Nix_ because we need the ability to evaluate old versions of Nixpkgs, as well as out-of-Nixpkgs code, thus enforcement will be a mess.

Nix 2.3 evaluating SRI hashes fine refutes neither objection. _What_ is the flow of argument you want me to follow?

Reply:
Is anyone here arguing for breaking non-SRI hashes?

Reply:
The current formulation sounds much more prescriptive than just "discouraging".  Point (1.) is a hard "require".  For (2.) I'm unsure about the meaning, maybe like excepting only fetchers that technically don't support.

Reply:
I'm open to changing it to "discourage", but... I don't know, if non-SRI hashes keep slipping in anyways because it's not required, what's the point of this? I guess that's why I chose that string wording.

Reply:
The point can be social pressure.
Pointing an authoritative document that "allows, but strongly discourages" old formats is better than endless discussions about "taste preferences".

The idea would be similar to the patched GCC from OpenBSD, that warns the user when the code uses dangerous functions like `strcpy()` or `gets()`. It does not block you from it but warns you.

Also, in my way for finding old format hashes, I have found buggy expressions, from dead projects to plain wrong files. Old hashes hide more than mere "backwards compatibility".

Reply:
So how does something like this sound: we strongly encourage using SRI hashes and `hash` in new packages, if there is no justifiable reason to not use them. Justifiable reasons include, but are not limited to, taking hashes directly from upstream, and using automated tooling that does not use SRI hashes.

I'm not sure about that last reason, but I also want to provide more than one example of a reason.

I'm also not sure about how to handle converting old packages, if at all. Thoughts welcome.

Reply:
Old packages can be converted by anyone interested ion converting them. I have started a preliminar work a long time ago :P

https://github.com/NixOS/nixpkgs/issues/176283

Further I believe this should be decided on a case by case basis. Maybe allowing old Nix tools is acceptable, whereas upstream hashes is not.

Reply:
> warns the user when the code uses dangerous functions

Not sure how apt this comparison is, as there's nothing fundamentally different about a sha256 hash in SRI format or one in base16 format.

> I'm open to changing it to "discourage", but... I don't know, if non-SRI hashes keep slipping in anyways because it's not required, what's the point of this? I guess that's why I chose that strong of a word.

Stale hashes is not something caused by a particular hash format. It happens because expressions bitrot and we'll have stale hashes in SRI format before long. Doing a mass conversion of hashes may turn up stale hashes, but doesn't have to since you would likely just convert them without refetching, since that is much quicker.

If you are concerned about stale hashes, I'd suggest trying to refetch fixed output derivations in nixpkgs with cache disabled. I believe this has been done quite a few times over the years and it should be easy to turn up useful tools for this.

> strongly encourage

Newer Nix versions already default to this, so there is already a lot of encouragement…

Reply:
"A lot of encouragement" that ends with "my version is too old" or "this is mere personal preferences, don't force them on me"? Wow cool.

Reply:
The version is probably too old for the format to be convenient to _produce_ but can evaluate the expressions with either hash format. For «mere personal preferences» argument new Nix making use of SRI format easier than non-SRI _is_ an encouragement, as submitters will typically use whatever is the easiest and changing the format is pointless churn so the initial hash format will stay.